File: Android/TRTCScenesDemo/trtcliveroom/src/main/java/com/tencent/liteav/liveroom/model/TRTCLiveRoom.java
package com.tencent.liteav.liveroom.model;

import android.annotation.TargetApi;
import android.content.Context;
import android.graphics.Bitmap;
import android.os.Handler;

import com.tencent.liteav.beauty.TXBeautyManager;
import com.tencent.liteav.liveroom.model.impl.TRTCLiveRoomImpl;
import com.tencent.rtmp.ui.TXCloudVideoView;

import java.util.List;

public abstract class TRTCLiveRoom {
    protected TRTCLiveRoom() {
    }

    /**
     * 获取 TRTCLiveRoom 单例对象
     *
     * @param context  Android 上下文，内部会转为 ApplicationContext 用于系统 API 调用
     * @return TRTCLiveRoom 实例
     *
     * @note 可以调用 {@link TRTCLiveRoom#destroySharedInstance()} 销毁单例对象
     */
    public static synchronized TRTCLiveRoom sharedInstance(Context context) {
        return TRTCLiveRoomImpl.sharedInstance(context);
    }

    /**
     * 销毁 TRTCLiveRoom 单例对象
     *
     * @note 销毁实例后，外部缓存的 TRTCLiveRoom 实例不能再使用，需要重新调用 {@link TRTCLiveRoom#sharedInstance(Context)} 获取新实例
     */
    public static void destroySharedInstance() {
        TRTCLiveRoomImpl.destroySharedInstance();
    }

    /**
     * 设置组件回调接口
     *
     * 您可以通过 TRTCLiveRoomDelegate 获得 TRTCLiveRoom 的各种状态通知
     *
     * @param delegate 回调接口
     *
     * @note 默认是在 Main Thread 中回调，如果需要自定义回调线程，可使用 {@link TRTCLiveRoom#setDelegateHandler(Handler)}
     */
    public abstract void setDelegate(TRTCLiveRoomDelegate delegate);

    /**
     * 设置组件回调的线程
     *
     * @param handler 线程
     */
    public abstract void setDelegateHandler(Handler handler);

    /**
     * 登录到组件系统
     */
    public abstract void login(int sdkAppId, String userId, String userSign, TRTCLiveRoomDef.TRTCLiveRoomConfig config, TRTCLiveRoomCallback.ActionCallback callback);

    /**
     * 退出登录
     */
    public abstract void logout(TRTCLiveRoomCallback.ActionCallback callback);

    /**
     * 设置用户信息
     * @param userName 用户昵称
     * @param avatarURL 用户头像
     * @param callback 设置结果回调
     */
    public abstract void setSelfProfile(String userName, String avatarURL, TRTCLiveRoomCallback.ActionCallback callback);

    /**
     * 创建房间（主播调用）
     *
     * 主播开播的正常调用流程是：
     * 1.【主播】调用 startCameraPreview() 打开摄像头预览，此时可以调整美颜参数。
     * 2.【主播】调用 createRoom 创建直播间，房间创建成功与否会通过 ActionCallback 通知给主播。
     *
     * @param roomId 房间标识。
     * @param roomInfo 房间信息，用于房间描述的信息，比如房间名称，封面信息。
     * @param callback 创建房间的结果回调
     */
    public abstract void createRoom(int roomId, TRTCLiveRoomDef.TRTCCreateRoomParam roomInfo, TRTCLiveRoomCallback.ActionCallback callback);

    /**
     * 销毁房间（主播调用）
     *
     * 主播在创建房间后，可以调用这个函数来销毁房间
     */
    public abstract void destroyRoom(TRTCLiveRoomCallback.ActionCallback callback);

    /**
     * 进入房间（观众调用）
     *
     * 观众观看直播的正常调用流程是：
     * 1.【观众】通过业务后台拿到最新的直播房间列表。
     * 2.【观众】选择一个直播间以后，调用 enterRoom() 进入该房间。
     *
     * @param roomId 房间标识
     * @param callback 进入房间的结果回调
     */
    public abstract void enterRoom(int roomId, TRTCLiveRoomCallback.ActionCallback callback);

    /**
     * 退出房间
     *
     * @param callback 退出房间的结果回调
     */
    public abstract void exitRoom(TRTCLiveRoomCallback.ActionCallback callback);

    /**
     * 获取房间列表的详细信息
     *
     * 通过业务后台拿到所有的房间号 roomIdList 后，调用该函数可以获取房间的详细信息
     *
     * @see {@link TRTCLiveRoomDef.TRTCLiveRoomInfo}
     *
     * @param roomIdList 房间号列表
     * @param callback 房间详细信息回调
     */
    public abstract void getRoomInfos(List<Integer> roomIdList, TRTCLiveRoomCallback.RoomInfoCallback callback);

    /**
     * 获取房间内所有的主播信息
     * @param callback 用户详细信息回调
     */
    public abstract void getAnchorList(TRTCLiveRoomCallback.UserListCallback callback);

    /**
     * 获取房间内所有的观众信息
     * @param callback 用户详细信息回调
     */
    public abstract void getAudienceList(TRTCLiveRoomCallback.UserListCallback callback);

    /**
     * 开启本地视频的预览画面
     *
     * @param isFront YES：前置摄像头；NO：后置摄像头。
     * @param view 承载视频画面的控件
     * @param callback 操作回调
     */
    public abstract void startCameraPreview(boolean isFront, TXCloudVideoView view, TRTCLiveRoomCallback.ActionCallback callback);

    /**
     * 停止本地视频采集及预览
     */
    public abstract void stopCameraPreview();

    /**
     * 开始直播
     * 1. 主播开播的时候调用
     * 2. 观众连麦后调用
     * @param streamId 用户直播的id
     * @param callback 操作回调
     */
    public abstract void startPublish(String streamId, TRTCLiveRoomCallback.ActionCallback callback);

    /**
     * 停止直播
     * 1. 主播停止直播时调用
     * 2. 观众退麦时调用
     * @param callback 操作回调
     */
    public abstract void stopPublish(TRTCLiveRoomCallback.ActionCallback callback);

    /**
     * 播放远端视频画面
     *
     * @param userId 对方的用户id
     * @param view 承载视频画面的控件
     * @param callback 操作回调
     *
     * @note 在 onAnchorEnter 回调时，调用这个接口
     */
    public abstract void startPlay(String userId, TXCloudVideoView view, TRTCLiveRoomCallback.ActionCallback callback);

    /**
     * 停止播放远端视频画面
     *
     * @param userId 对方的用户信息
     * @param callback 操作回调
     *
     * @note 在 onAnchorExit 回调时，调用这个接口
     */
    public abstract void stopPlay(String userId, TRTCLiveRoomCallback.ActionCallback callback);

    /**
     * 观众请求连麦
     *
     * 主播和观众的连麦流程可以简单描述为如下几个步骤：
     * 1. 【观众】调用 requestJoinAnchor() 向主播发起连麦请求。
     * 2. 【主播】会收到 {@link TRTCLiveRoomDelegate#onRequestJoinAnchor} 的回调通知。
     * 3. 【主播】调用 responseJoinAnchor() 确定是否接受观众的连麦请求。
     * 4. 【观众】会收到 {@link TRTCLiveRoomCallback.ActionCallback} 回调通知，可以得知请求是否被同意。
     * 5. 【观众】如果请求被同意，则调用 startCameraPreview() 开启本地摄像头。
     * 6. 【观众】然后调用 startPublish() 正式进入推流状态。
     * 7. 【主播】一旦观众进入连麦状态，主播就会收到 {@link TRTCLiveRoomDelegate#onAnchorEnter)} 通知。
     * 8. 【主播】主播调用 startPlay() 就可以看到连麦观众的视频画面。
     * 9. 【观众】如果直播间里已经有其他观众正在跟主播进行连麦，那么新加入的这位连麦观众也会收到 onAnchorEnter() 通知，用于展示（startPlay）其他连麦者的视频画面。
     *
     * @param reason 连麦原因
     * @param callback 请求连麦的回调
     *
     * @see {@link TRTCLiveRoomDelegate#onRequestJoinAnchor}
     */
    public abstract void requestJoinAnchor(String reason, TRTCLiveRoomCallback.ActionCallback callback);

    /**
     * 主播处理连麦请求
     *
     * 主播在收到 {@link TRTCLiveRoomDelegate#onRequestJoinAnchor} 回调之后会需要调用此接口来处理观众的连麦请求。
     *
     * @param userId 观众 ID
     * @param agree true：同意；false：拒绝
     * @param reason 同意/拒绝连麦的原因描述
     */
    public abstract void responseJoinAnchor(String userId, boolean agree, String reason);

    /**
     * 主播踢除连麦观众
     *
     * 主播调用此接口踢除连麦观众后，被踢连麦观众会收到 {@link TRTCLiveRoomDelegate#onKickoutJoinAnchor()} 回调通知
     *
     * @param userId 连麦观众 ID
     *
     * @see {@link TRTCLiveRoomDelegate#onKickoutJoinAnchor()}
     */
    public abstract void kickoutJoinAnchor(String userId, TRTCLiveRoomCallback.ActionCallback callback);

    /**
     * 请求跨房 PK
     *
     * 主播和主播之间可以跨房间 PK，两个正在直播中的主播 A 和 B，他们之间的跨房 PK 流程如下：
     * 1. 【主播 A】调用 requestRoomPK() 向主播 B 发起连麦请求。
     * 2. 【主播 B】会收到 {@link TRTCLiveRoomDelegate#onRequestRoomPK} 回调通知。
     * 3. 【主播 B】调用 responseRoomPK() 确定是否接受主播 A 的 PK 请求。
     * 4. 【主播 B】如果接受了主播 A 的要求，等待 {@link TRTCLiveRoomDelegate#onAnchorEnter)} 通知，然后调用 startPlay() 来显示主播 A 的视频画面。
     * 5. 【主播 A】会收到 {@link TRTCLiveRoomCallback.ActionCallback} 回调通知，可以得知请求是否被同意。
     * 6. 【主播 A】如果请求被同意，等待 {@link TRTCLiveRoomDelegate#onAnchorEnter)} 通知，然后调用 startPlay() 来显示主播 B 的视频画面
     *
     * @param roomId 被邀约房间 ID
     * @param userId 被邀约主播 ID
     * @param callback 请求跨房 PK 的结果回调
     *
     * @see {@link TRTCLiveRoomDelegate#onRequestRoomPK}
     */
    public abstract void requestRoomPK(int roomId, String userId, TRTCLiveRoomCallback.ActionCallback callback);

    /**
     * 响应跨房 PK 请求
     *
     * 主播响应其他房间主播的 PK 请求。
     *
     * @param userId 发起 PK 请求的主播 ID
     * @param agree true：同意；false：拒绝
     * @param reason 同意/拒绝 PK 的原因描述
     */
    public abstract void responseRoomPK(String userId, boolean agree, String reason);

    /**
     * 退出跨房 PK
     *
     * 当两个主播中的任何一个退出跨房 PK 状态后，另一个主播会收到 {@link TRTCLiveRoomDelegate#onQuitRoomPK} 回调通知。
     *
     * @param callback 退出跨房 PK 的结果回调
     */
    public abstract void quitRoomPK(TRTCLiveRoomCallback.ActionCallback callback);

    /**
     * 切换前后摄像头
     */
    public abstract void switchCamera();

    /**
     * 设置是否镜像展示
     */
    public abstract void setMirror(boolean isMirror);

    /**
     * 静音本地音频
     */
    public abstract void muteLocalAudio(boolean mute);

    /**
     * 静音远端音频
     */
    public abstract void muteRemoteAudio(String userId, boolean mute);

    /**
     * 静音所有远端音频
     */
    public abstract void muteAllRemoteAudio(boolean mute);

    /**
     * 发送文本消息
     *
     * @param message 文本消息
     * @param callback 发送结果回调
     */
    public abstract void sendRoomTextMsg(String message, TRTCLiveRoomCallback.ActionCallback callback);

    /**
     * 发送自定义文本消息
     *
     * @param cmd 命令字，由开发者自定义，主要用于区分不同消息类型
     * @param message 文本消息
     * @param callback 发送结果回调
     */
    public abstract void sendRoomCustomMsg(String cmd, String message, TRTCLiveRoomCallback.ActionCallback callback);

    /**
     * 是否在界面中展示debug信息
     */
    public abstract void showVideoDebugLog(boolean isShow);

    /**
     * BGM控制相关
     */
    public abstract TRTCBGMManager getBGMManger();

    /**
     * 美颜相关
     */
    public abstract TXBeautyManager getBeautyManager();

    /**
     * 设置指定素材滤镜特效
     *
     * @param image 指定素材，即颜色查找表图片。**必须使用 png 格式**
     */
    public abstract void setFilter(Bitmap image);

    /**
     * 设置滤镜浓度
     * <p>
     * 在美女秀场等应用场景里，滤镜浓度的要求会比较高，以便更加突显主播的差异。
     * 我们默认的滤镜浓度是0.5，如果您觉得滤镜效果不明显，可以使用下面的接口进行调节。
     *
     * @param concentration 从0到1，越大滤镜效果越明显，默认值为0.5。
     */
    public abstract void setFilterConcentration(float concentration);

    /**
     * 设置绿幕背景视频（企业版有效，其它版本设置此参数无效）
     * <p>
     * 此处的绿幕功能并非智能抠背，它需要被拍摄者的背后有一块绿色的幕布来辅助产生特效
     *
     * @param file 视频文件路径。支持 MP4；null：表示关闭特效。
     */
    @TargetApi(18)
    public abstract void setGreenScreenFile(String file);
}

File: iOS/TRTCScenesDemo/TRTCScenesDemo/TRTCLiveRoomDemo/model/TRTCLiveRoom.swift
//
//  TRTCLiveRoom.swift
//  TRTCLiveRoom
//
//  Created by Xiaoya Liu on 2020/2/7.
//  Copyright © 2020 Xiaoya Liu. All rights reserved.
//

import UIKit

@objc public protocol TRTCLiveRoom: class {
    typealias Callback = (_ code: Int, _ message: String?) -> Void
    typealias ResponseCallback = (_ agreed: Bool, _ reason: String?) -> Void
    typealias RoomInfoCallback = (_ code: Int, _ message: String?, _ roomList: [TRTCLiveRoomInfo]) -> Void
    typealias UserListCallback = (_ code: Int, _ message: String, _ userList: [TRTCLiveUserInfo]) -> Void
    
    /// 登录到组件系统
    /// - Parameters:
    ///   - sdkAppID: TRTC SDKAppID
    ///   - userID: 用户ID
    ///   - userSig: 用户签名
    ///   - config: liveRoom的配置，登录时初始化，之后不可变更
    ///   - callback: 登录回调
    /// - Note:
    ///   - userSig 建议设定 7 天，能够有效规避 usersign 过期导致的 IM 收发消息失败、TRTC 连麦失败等情况
    @objc func login(sdkAppID: Int, userID: String, userSig: String, config: TRTCLiveRoomConfig, callback: Callback?)
    
    /// 用户登出
    /// - Parameter callback: 登出回调
    @objc func logout(_ callback: Callback?)
    
    /// 设置用户信息
    /// - Parameters:
    ///   - name: 用户昵称
    ///   - avatarURL: 用户头像地址
    ///   - callback: 设置回调
    @objc func setSelfProfile(name: String, avatarURL: String?, callback: Callback?)
    
    /// 主播创建房间
    /// 主播开播的正常调用流程是：
    /// 1.【主播】调用 startCameraPreview() 打开摄像头预览，此时可以调整美颜参数。
    /// 2.【主播】调用 createRoom() 创建直播间，房间创建成功与否会通过 TRTCLiveRoomDelegate 通知给主播。
    /// 3.【主播】调用 startPublish() 开始推流
    /// - Parameters:
    ///   - roomID: 房间ID
    ///   - roomInfo: 房间参数
    ///   - callback: 创建房间结果回调
    /// - Note:
    ///   - 主播开始直播的时候调用，可重复创建自己已创建过的房间。
    @objc func createRoom(roomID: UInt32, roomParam: TRTCCreateRoomParam, callback: Callback?)
    
    /// 主播销毁房间
    /// - Parameter callback: 销毁房间回调
    /// - Note:
    ///   - 主播结束直播的时候调用
    @objc func destroyRoom(callback: Callback?)
    
    /// 观众进入房间
    ///观众观看直播的正常调用流程是：
    ///1.【观众】通过业务后台拿到最新的直播房间列表。
    ///2.【观众】选择一个直播间以后，调用 enterRoom() 进入该房间。
    ///3.【观众】调用 startPlay() 播放主播的画面。
    /// - Parameters:
    ///   - roomID: 房间ID
    ///   - callback: 进入房间回调
    /// - Note:
    ///   - 观众进入直播房间的时候调用
    ///   - 主播不可调用这个接口进入自己已创建的房间，而要用createRoom
    @objc func enterRoom(roomID: UInt32, callback: Callback?)
    
    /// 观众离开房间
    /// - Parameter callback: 离开房间回调
    /// - Note:
    ///   - 观众离开直播房间的时候调用
    ///   - 主播不可调用这个接口离开房间
    @objc func exitRoom(callback: Callback?)
    
    /// 获取房间信息
    /// - Parameter roomIDs: 房间ID列表
    /// - Parameter callback: 房间信息回调
    @objc func getRoomInfos(roomIDs: [UInt32], callback: RoomInfoCallback?)
    
    /// 获取主播列表
    /// - Parameter callback: 主播列表回调
    @objc func getAnchorList(callback: UserListCallback?)
    
    /// 获取主播列表
    /// - Parameter callback: 观众列表回调
    @objc func getAudienceList(callback: UserListCallback?)
    
    /// 开启视频预览
    /// - Parameters:
    ///   - frontCamera: 使用前置摄像头
    ///   - view: 用来渲染主播画面的 View
    ///   - callback: 开启预览回调
    @objc func startCameraPreview(frontCamera: Bool, view: UIView, callback: Callback?)
    
    /// 关闭视频预览
    @objc func stopCameraPreview()
    
    /// 开启推流
    /// - Parameters:
    ///   - streamID: 自定义流ID，传空时系统会生成默认的流ID。这里也可以传入完整的 CDN 地址，用于转推流
    ///   - callback: 开启推流回调
    @objc func startPublish(streamID: String?, callback: Callback?)
    
    /// 关闭推流
    /// - Parameter callback: 关闭推流回调
    @objc func stopPublish(callback: Callback?)
    
    /// 播放主播画面
    /// - Parameters:
    ///   - userID: 主播用户ID
    ///   - view: 渲染画面用的view
    ///   - callback: 播放回调
    @objc func startPlay(userID: String, view: UIView, callback: Callback?)
    
    /// 停止播放主播画面
    /// - Parameters:
    ///   - userID: 主播用户ID
    ///   - callback: 停止播放回调
    @objc func stopPlay(userID: String, callback: Callback?)
    
    /**
     * 观众请求连麦
     *
     * 主播和观众的连麦流程可以简单描述为如下几个步骤：
     * 1. 【观众】调用 requestJoinAnchor() 向主播发起连麦请求。
     * 2. 【主播】会收到 TRTCLiveRoomDelegate onRequestJoinAnchor 的回调通知。
     * 3. 【主播】调用 responseJoinAnchor() 确定是否接受观众的连麦请求。
     * 4. 【观众】会收到 responseCallback  回调通知，可以得知请求是否被同意。
     * 5. 【观众】如果请求被同意，则调用 startCameraPreview() 开启本地摄像头。
     * 6. 【观众】然后调用 startPublish() 正式进入推流状态。
     * 7. 【主播】一旦观众进入连麦状态，主播就会收到 TRTCLiveRoomDelegate onAnchorEnter 通知。
     * 8. 【主播】主播调用 startPlay() 就可以看到连麦观众的视频画面。
     * 9. 【观众】如果直播间里已经有其他观众正在跟主播进行连麦，那么新加入的这位连麦观众也会收到 onAnchorEnter() 通知，用于展示（startPlay）其他连麦者的视频画面。
 * */
    
    /// 观众端请求连麦
    /// - Parameters:
    ///   - reason: 连麦请求原因
    ///   - responseCallback: 主播是否同意连麦的回调
    ///   - callback: 请求错误回调
    /// - Note: 观众发起请求后，主播端会收到`onRequestJoinAnchor`回调
    @objc func requestJoinAnchor(reason: String?, responseCallback: ResponseCallback?, callback: Callback?)
    
    /// 主播回复观众连麦请求
    /// - Parameters:
    ///   - user: 观众
    ///   - agree: 是否同意
    ///   - reason: 回复原因
    /// - Note: 主播回复后，观众端会收到`requestJoinAnchor`传入的`responseCallback`回调
    @objc func responseJoinAnchor(userID: String, agree: Bool, reason: String?)
    
    /// 主播将连麦观众下麦
    /// - Parameters:
    ///   - userID: 观众ID
    ///   - callback: 下麦回调
    /// - Note: 下麦后，被下麦的观众会收到`trtcLiveRoomOnKickoutJoinAnchor`回调
    @objc func kickoutJoinAnchor(userID: String, callback: Callback?)
    
    /**
        * 请求跨房 PK
        *
        * 主播和主播之间可以跨房间 PK，两个正在直播中的主播 A 和 B，他们之间的跨房 PK 流程如下：
        * 1. 【主播 A】调用 requestRoomPK() 向主播 B 发起连麦请求。
        * 2. 【主播 B】会收到 TRTCLiveRoomDelegate onRequestRoomPK 回调通知。
        * 3. 【主播 B】调用 responseRoomPK() 确定是否接受主播 A 的 PK 请求。
        * 4. 【主播 B】如果接受了主播 A 的要求，等待 TRTCLiveRoomDelegate onAnchorEnter 通知，然后调用 startPlay() 来显示主播 A 的视频画面。
        * 5. 【主播 A】会收到 responseCallback 回调通知，可以得知请求是否被同意。
        * 6. 【主播 A】如果请求被同意，等待 TRTCLiveRoomDelegate onAnchorEnter 通知，然后调用 startPlay() 来显示主播 B 的视频画面
        */
    
    /// 主播向另一个主播请求连麦PK
    /// - Parameters:
    ///   - roomID: 对方主播房间ID
    ///   - userID: 对方主播用户ID
    ///   - responseCallback: 对方主播是否同意跨房PK的回调
    ///   - callback: 跨房PK错误回调
    /// - Note: 发起请求后，对方主播会收到`onRequestRoomPK`回调
    @objc func requestRoomPK(roomID: UInt32, userID: String, responseCallback: ResponseCallback?, callback: Callback?)
    
    /// 主播回复跨房PK的请求
    /// - Parameters:
    ///   - user: 对方主播
    ///   - agree: 是否同意
    ///   - reason: 回复原因    /// - Note: 主播回复后，对方主播会收到`requestRoomPK`传入的`responseCallback`回调
    @objc func responseRoomPK(userID: String, agree: Bool, reason: String?)
    
    /// 主播退出连麦
    /// - Parameter callback: 退出连麦的回调
    /// - Note: 退出后，对方主播会收到`trtcLiveRoomOnQuitRoomPK`回调
    @objc func quitRoomPK(callback: Callback?)
    
    /// 切换前后摄像头
    @objc func switchCamera()
    
    /// 切换视频画面镜像
    /// - Parameter isMirror: 开启镜像
    @objc func setMirror(_ isMirror: Bool)
    
    /// 将麦克风静音
    /// - Parameter isMuted: 开启静音
    @objc func muteLocalAudio(_ isMuted: Bool)
    
    /// 将某个主播的声音静音
    /// - Parameters:
    ///   - userID: 主播ID
    ///   - isMuted: 开启静音
    @objc func muteRemoteAudio(userID: String, isMuted: Bool)
    
    /// 将所有主播的声音静音
    /// - Parameter isMuted: 开启静音
    @objc func muteAllRemoteAudio(_ isMuted: Bool)
    
    /// 发送文本消息，房间内所有成员都可见
    /// - Parameters:
    ///   - message: 文本消息
    ///   - callback: 发送回调
    @objc func sendRoomTextMsg(message: String, callback: Callback?)
    
    /// 发送自定义消息
    /// - Parameters:
    ///   - command: 命令字，由开发者自定义，主要用于区分不同消息类型
    ///   - message: 本文消息
    ///   - callback: 发送回调
    @objc func sendRoomCustomMsg(command: String, message: String, callback: Callback?)
    
    /// 开启渲染页面上的 Debug 信息
    /// - Parameter isShow: 开启 Debug 信息显示
    @objc func showVideoDebugLog(_ isShow: Bool)
    
    /* 美颜及动效参数管理
    *
    * 通过美颜管理，您可以使用以下功能：
    * - 设置"美颜风格"、“美白”、“红润”、“大眼”、“瘦脸”、“V脸”、“下巴”、“短脸”、“小鼻”、“亮眼”、“白牙”、“祛眼袋”、“祛皱纹”、“祛法令纹”等美容效果。
    * - 调整“发际线”、“眼间距”、“眼角”、“嘴形”、“鼻翼”、“鼻子位置”、“嘴唇厚度”、“脸型”
    * - 设置人脸挂件（素材）等动态效果
    * - 添加美妆
    * - 进行手势识别
    */
    @objc func getBeautyManager() -> TXBeautyManager
    
    /**
     * 设置指定素材滤镜特效
     *
     * @param image 指定素材，即颜色查找表图片。**必须使用 png 格式**
     */
    @objc func setFilter(image: UIImage)

    /**
     * 设置滤镜浓度
     *
     * 在美女秀场等应用场景里，滤镜浓度的要求会比较高，以便更加突显主播的差异。
     * 我们默认的滤镜浓度是0.5，如果您觉得滤镜效果不明显，可以使用下面的接口进行调节。
     *
     * @param concentration 从0到1，越大滤镜效果越明显，默认值为0.5。
     */
    @objc func setFilterConcentration(concentration: Float)

    /**
     * 设置绿幕背景视频（企业版有效，其它版本设置此参数无效）
     *
     * 此处的绿幕功能并非智能抠背，需要被拍摄者的背后有一块绿色的幕布来辅助产生特效
     *
     * @param file 视频文件路径。支持 MP4; nil 表示关闭特效。
     */
    @objc func setGreenScreenFile(file: URL?)
    
    /// 背景音乐管理实例
    @objc func getAudioEffectManager() -> TRTCAudioEffectManagerImpl
}
